internal/entitycache/cache.go

```diff
@@ -132,8 +131,8 @@ func (c *Cache) Entity(id *charm.URL, fields ...string) (*mongodoc.Entity, error
 // BaseEntity returns the base entity with the given id. If the entity is not
 // found, it returns an error with a params.ErrNotFound cause.
 // The returned entity will have at least the given fields filled out.
-func (c *Cache) BaseEntity(id *charm.URL, fields ...string) (*mongodoc.BaseEntity, error) {
-	e, err := c.baseEntities.entity(mongodoc.BaseURL(id), fields...)
+func (c *Cache) BaseEntity(id *charm.URL, fields map[string]int) (*mongodoc.BaseEntity, error) {
+	e, err := c.baseEntities.entity(mongodoc.BaseURL(id), fields)
```

Comment:
This is not a sufficiently correct way to calculate the correct base URL for an entity.

Imagine there are two charms

~charmers/trusty/mysql-1 -> trusty/mysql-0
~rogpeppe/trusty/mysql-1 -> trusty/mysql-1

It is vital that the base url for trusty/mysql-0 is calculated to be ~charmers/mysql and trusty/mysql-1 is calculated to be ~rogpeppe/trusty/mysql-1. That can only be worked out from a database query, not by manipulating the URL.